scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java

```diff
@@ -766,6 +770,211 @@ public DatasetId storeSimulationResults(final PlanMetadata planMetadata,
+  {
+    final var simulatedActivities = new HashMap<SimulatedActivityId, SimulatedActivity>();
+    for(final var simulatedActivityJson: simulatedActivitiesArray) {
+      //if no duration, this is an unfinished activity
+      if(simulatedActivityJson.asJsonObject().isNull("duration")) continue;
+      final var activityDuration = GraphQLParsers.durationP.parse(simulatedActivityJson.asJsonObject().get("duration")).getSuccessOrThrow();
+      final var activityId = simulatedActivityJson.asJsonObject().getJsonNumber("id").longValue();
+      SimulatedActivityId parentId = null;
+      if(!simulatedActivityJson.asJsonObject().isNull("parent_id")){
+        parentId = new SimulatedActivityId(simulatedActivityJson.asJsonObject().getJsonNumber("parent_id").longValue());
+      }
+      final var startOffset = instantFromStart(simulationStart,durationFromPGInterval(simulatedActivityJson.asJsonObject().getString("start_offset")));
+      final var computedAttributes = serializedValueP.parse(simulatedActivityJson.asJsonObject().get("attributes")).getSuccessOrThrow();
+      final var activityDirective = simulatedActivityJson.asJsonObject().getJsonObject("activity_directive");
+      final var activityDirectiveId = new ActivityDirectiveId(activityDirective.getInt("id"));
+      final var activityDirectiveArguments = activityDirective.getJsonObject("arguments");
+      final var deserializedArguments = BasicParsers
+          .mapP(serializedValueP)
+          .parse(activityDirectiveArguments)
+          .getSuccessOrThrow((reason) -> new InvalidJsonException(new InvalidEntityException(List.of(reason))));
+      final var activityType = activityDirective.getString("type");
+      final var simulatedActivity = new SimulatedActivity(
+          activityType,
+          deserializedArguments,
+          startOffset,
+          activityDuration,
+          parentId,
+          List.of(),
+          Optional.of(activityDirectiveId),
+          computedAttributes
+      );
+      simulatedActivities.put(new SimulatedActivityId(activityId), simulatedActivity);
+    }
+    return simulatedActivities;
+  }
+
+  /**
+   * Returns the simulation dataset id if the simulation
+   * - covers the entire planning horizon
+   * - corresponds to the plan revision
+   * @param planMetadata the plan metadata containing the planning horizon and plan revision
+   * @return optionally a simulation dataset id
+   */
+  private Optional<SimulationDatasetId> getSuitableSimulationResults(PlanMetadata planMetadata) throws PlanServiceException, IOException {
+    final var request = """
+        {
+          simulation(where: {plan_id: {_eq: %d}}) {
+            id
+            simulation_datasets(where: {plan_revision: {_eq: %d}, canceled: {_eq: false}}) {
+              id
+              simulation_start_time
+              simulation_end_time
+            }
+          }
+        }""".formatted(planMetadata.planId().id(),
+                       planMetadata.planRev());
```

Comment:
Proposal: The following query more directly accesses the desired simulation dataset. It removes the unused `id` from `simulation` that's fetched, replacing it with `arguments`, which can be used to ensure that the args that the sim dataset used matches the one that the scheduler will use (although if that information's already accessible in this method, that part of the query can be removed). It additionally moves the sim bounds check into the `where` clause of the query and only queries successful sim runs to avoid fetching unnecessary rows.

```java
    final var request =
      """
      {
        simulation_dataset(
          where: {
            canceled: {_eq: false},
            status: {_eq: "success"},
            plan_revision: {_eq: %d},
            simulation_start_time: {_eq: "%s"},
            simulation_end_time: {_eq: "%s"},
            simulation: {plan_id: {_eq: %d}}
          }) {
            id
            arguments
            simulation {
              arguments
            }
          }
      }""".formatted(
              planMetadata.planRev(),
              planMetadata.horizon().getStartInstant(), // this may need a parse
              planMetadata.horizon().getEndInstant(), // this may need a parse 
              planMetadata.planId().id());
```

Reply:
I will use that, thank you.